pymodbus/transport/transport.py

```diff
@@ -310,12 +311,26 @@ def data_received(self, data: bytes):
         self.datagram_received(data, None)
 
     def datagram_received(self, data: bytes, addr: tuple):
-        """Receive datagram (UDP connections)."""
+        """Receive datagram"""
         if self.comm_params.handle_local_echo and self.sent_buffer:
-            Log.debug("recv skipping (local_echo): {} addr={}", data, ":hex", addr)
-            if self.sent_buffer in data:
-                data, self.sent_buffer = data.replace(self.sent_buffer, b"", 1), b""
+            if data.startswith(self.sent_buffer):
+                Log.debug(
+                    "recv skipping (local_echo): {} addr={}",
+                    self.sent_buffer,
+                    ":hex",
+                    addr,
+                )
+                data = data[len(self.sent_buffer) :]
+                self.sent_buffer = b""
+                self.wait_for_local_echo.set_result(True)
+            elif self.sent_buffer.startswith(data):
```

Comment:
If that happens then it is not a local echo.

Reply:
How so? So a partial read of a local echo is not a local echo??

This path is used by the serial port and it is definitely possible to have a partial read of the data.

Reply:
Of course, but it looks suspect to me, maybe because it is all mixed with the future.